File: backend/src/root.cpp
#include <gstreamermm/tee.h>

#include "root.hpp"

#include <iostream>

using namespace Ats;
using namespace std;

Root::Root (const Glib::RefPtr<Gst::Bin> bin,
            const Metadata& m, const Settings& s,
            std::shared_ptr<Video_data> vd, std::shared_ptr<Audio_data> ad) : _settings(s) {
    uint stream = m.stream;
    
    _bin   = bin;
    _src   = Gst::ElementFactory::create_element("udpsrc");
    auto parse = Gst::ElementFactory::create_element("tsparse");
    _tee   = Gst::Tee::create();
    _video_sender = vd;
    _audio_sender = ad;

    _src->set_property("uri", m.uri);
    _src->set_property("buffer-size", 2147483647);

    _bin->add(_src)->add(parse)->add(_tee);

    _src->link(parse)->link(_tee);

    m.for_analyzable ([this,stream](const Meta_channel& c) { build_cb(stream,c); });
}

Root::~Root() {
    _bin.reset();
}

unique_ptr<Root>
Root::create (const Glib::RefPtr<Gst::Bin> bin,
              const Metadata& m, const Settings& s,
              std::shared_ptr<Video_data> vd, std::shared_ptr<Audio_data> ad) {
    if (! m.to_be_analyzed()) return unique_ptr<Root>(nullptr);
    else return unique_ptr<Root>(new Root(bin, m, s, vd, ad));
}

void
Root::apply (const Settings& s) {
    _settings = s;
    for (auto& branch : _branches) {
        branch->apply(s);
    }
}

void
Root::build_cb (const uint stream, const Meta_channel& c) {
    string demux_name = "demux_";
    demux_name += std::to_string(stream);
    demux_name += "_";
    demux_name += std::to_string(c.number);
    //cout << "Demux name: " << demux_name << endl;

    auto queue = Gst::ElementFactory::create_element("queue");
    auto _demux = Gst::ElementFactory::create_element("tsdemux",demux_name);
    
    _demux->set_property("program-number", c.number);
    queue->set_property("max-size-buffers", 200000);
    queue->set_property("max-size-bytes", 429496729);
    
    auto sinkpad = queue->get_static_pad("sink"); 
    auto srcpad = _tee->get_request_pad("src_%u");

    _bin->add(queue)->add(_demux);
    queue->link(_demux);

    srcpad->link(sinkpad);
    // GST_DEBUG_BIN_TO_DOT_FILE(GST_BIN(_bin->gobj()), GST_DEBUG_GRAPH_SHOW_ALL, "pipeline");
    _demux->signal_pad_added().connect([this, stream, c](const Glib::RefPtr<Gst::Pad>& p)
                                       { build_branch(stream, c, p); });
    //_demux->signal_pad_removed().connect([this, stream, num](const Glib::RefPtr<Gst::Pad>& p)
    //				      { destroy_branch(stream, num, p); });
}

void
Root::build_branch (const uint stream,
		    const Meta_channel c,
		    const Glib::RefPtr<Gst::Pad>& p) {
    auto num   = c.number;
    auto pname = p->get_name();
    auto pcaps = p->get_current_caps()->get_structure(0).get_name();
    vector<Glib::ustring> name_toks = Glib::Regex::split_simple("_", pname);
    vector<Glib::ustring> caps_toks = Glib::Regex::split_simple("/", pcaps);

    auto& type = caps_toks[0];	
		    
    if (type != "video" && type != "audio") return;
    
    auto pid  = strtoul(name_toks[2].data(), NULL, 16);

    if (! c.find_pid(pid)->to_be_analyzed) return;

    auto branch = Branch::create(type, stream, num, pid, _video_sender, _audio_sender);

    if (branch == nullptr) return;

    branch->add_to_pipe(_bin);
    branch->plug(p);
	
    branch->signal_pad_added().connect([this](std::shared_ptr<Pad> p){
            std::scoped_lock lock{_mutex_pad};
            _pad_added.emit(p);
        });
    
    if (branch->type() == Branch::Type::Audio) {
        Audio_branch* b = dynamic_cast<Audio_branch*>(branch.get());
        b->signal_audio_pad_added().connect([this](std::shared_ptr<Pad> p){
                std::scoped_lock lock{_mutex_audio_pad};
                _audio_pad_added.emit(p);
            });
    }

    branch->apply(_settings);
    _branches.push_back(std::move(branch));
    _bin->set_state(Gst::State::STATE_PLAYING);
}

void
Root::destroy_branch (const uint stream,
                      const uint num,
                      const Glib::RefPtr<Gst::Pad>& p) {
	
    auto pname = p->get_name();
    auto pcaps = p->get_current_caps()->get_structure(0).get_name();

    vector<Glib::ustring> name_toks = Glib::Regex::split_simple("_", pname);
    vector<Glib::ustring> caps_toks = Glib::Regex::split_simple("/", pcaps);

    auto& type = caps_toks[0];	
		    
    if (type != "video" && type != "audio") return;
    
    auto pid  = strtoul(name_toks[2].data(), NULL, 16);

    for (auto it = _branches.begin(); it != _branches.end(); it++) {
        if (*(*it) == make_tuple(stream, num, pid)) {
            _branches.erase(it);
            break;
        }
    }   
}

File: backend/src/root.rs
use gst::prelude::*;
use gst;
use std::sync::{Arc,Mutex,Weak};
use std::sync::mpsc::Sender;
use metadata::{Channel,Structure};
use settings::Settings;
use signals::Signal;
use pad::SrcPad;
use branch::{Typ,Branch};
use std::u32;

pub struct Root {
    settings:      Arc<Mutex<Option<Settings>>>,
    branches:      Arc<Mutex<Vec<Branch>>>,
    pub pad_added: Arc<Mutex<Signal<SrcPad>>>,
    pub audio_pad_added: Arc<Mutex<Signal<SrcPad>>>,
}

impl Root {

    fn build_branch (branches: &mut Vec<Branch>,
                     stream: String, chan: &Channel,
                     settings: Option<Settings>,
                     added: Arc<Mutex<Signal<SrcPad>>>,
                     audio_added: Arc<Mutex<Signal<SrcPad>>>,
                     bin: &gst::Pipeline,
                     pad: &gst::Pad,
                     sender_data: Weak<Mutex<Sender<(Typ,String,u32,u32,gst::Buffer)>>>,
                     sender_status: Weak<Mutex<Sender<(String,u32,u32,bool)>>>) {
        let pname = pad.get_name();
        let pcaps = String::from(pad.get_current_caps().unwrap().get_structure(0).unwrap().get_name());
        let name_toks: Vec<&str> = pname.split('_').collect();
        let caps_toks: Vec<&str> = pcaps.split('/').collect();
        
        if name_toks.len() != 3 || caps_toks.is_empty() { return };
        let typ = caps_toks[0];
        let pid = u32::from_str_radix(name_toks[2], 16).unwrap();
        
        /* No corresponding pid in config */
        if let None = chan.find_pid(pid){
            return;
        };

        debug!("Root::build_branch [{}]", pid);

        if let Some(branch) = Branch::new(stream,
                                          chan.number,
                                          pid,
                                          typ,
                                          settings,
                                          sender_data,
                                          sender_status) {
            branch.add_to_pipe(&bin);
            branch.plug(&pad);
            match branch {
                Branch::Video(ref b) => b.pad_added.lock().unwrap().connect(move |p| {
                    added.lock().unwrap().emit(p)
                }),
                Branch::Audio(ref b) => {
                    b.pad_added.lock().unwrap().connect(move |p| {
                        added.lock().unwrap().emit(p)
                    });
                    b.audio_pad_added.lock().unwrap().connect(move |p| {
                        audio_added.lock().unwrap().emit(p)
                    });
                },
            };

            debug!("Root::build_branch [{} ready]", pid);
            branches.push(branch);
        }
    }
    
    pub fn new(bin: &gst::Pipeline,
               m: &Structure,
               settings: Option<Settings>,
               sender_data: &Arc<Mutex<Sender<(Typ,String,u32,u32,gst::Buffer)>>>,
               sender_status: &Arc<Mutex<Sender<(String,u32,u32,bool)>>>)
               -> Option<Root> {
        debug!("Root::new");

        let src             = gst::ElementFactory::make("udpsrc", None).unwrap();
        let tee             = gst::ElementFactory::make("tee", None).unwrap();
        let settings        = Arc::new(Mutex::new(settings));
        let branches        = Arc::new(Mutex::new(Vec::new()));
        let pad_added       = Arc::new(Mutex::new(Signal::new()));
        let audio_pad_added = Arc::new(Mutex::new(Signal::new()));
        
        src.set_property("uri", &m.uri).unwrap();
        src.set_property("buffer-size", &2_147_483_647).unwrap();

        bin.add_many(&[&src, &tee]).unwrap();
        src.link(&tee).unwrap();

        let id = m.id.clone();

        for chan in &m.channels {
            let demux_name = format!("demux_{}_{}_{}_{}",
                                     id.clone(), chan.number, chan.service_name, chan.provider_name);

            let queue = gst::ElementFactory::make("queue", None).unwrap();
            let demux = gst::ElementFactory::make("tsdemux", Some(demux_name.as_str())).unwrap();

            demux.set_property("program-number", &(chan.number as i32)).unwrap();
            queue.set_property("max-size-time", &0u64).unwrap();
            queue.set_property("max-size-buffers", &0u32).unwrap();
            queue.set_property("max-size-bytes", &0u32).unwrap();

            let sinkpad = queue.get_static_pad("sink").unwrap();
            let srcpad  = tee.get_request_pad("src_%u").unwrap();

            bin.add_many(&[&queue,&demux]).unwrap();
            queue.link(&demux).unwrap();

            // TODO check
            let _ = srcpad.link(&sinkpad);
            
            let stream   = id.clone();
            let chan     = chan.clone();
            let bin_weak = bin.downgrade();
            let settings_c = settings.clone();
            let branches_weak = branches.clone();
            let pad_added_c = pad_added.clone();
            let audio_pad_added_c = audio_pad_added.clone();
            
            let sender_data = Arc::downgrade(sender_data);
            let sender_status = Arc::downgrade(sender_status);
            
            demux.connect_pad_added(move | _, pad | {
                let bin      = bin_weak.clone().upgrade().unwrap();
                //let branches = branches_weak.upgrade().unwrap();
                let settings = settings_c.lock().unwrap();
                Root::build_branch(&mut branches_weak.lock().unwrap(),
                                   stream.clone(), &chan,
                                   *settings,
                                   pad_added_c.clone(), audio_pad_added_c.clone(),
                                   &bin,
                                   pad,
                                   sender_data.clone(),
                                   sender_status.clone());
            });
        };

        Some(Root { settings, branches, pad_added, audio_pad_added })
    }

    pub fn apply_settings(&mut self, s: Settings) {
        self.branches.lock()
            .unwrap()
            .iter_mut()
            .for_each(|b : &mut Branch| b.apply_settings(s));
        *self.settings.lock().unwrap() = Some(s);
    }
}
